### DISCUSSION TITLE
should systems be named `*_system`

### DISCUSSION THREAD

**Discussion by mockersf** [bevyengine/bevy#2804] [General]
While I have an opinion on the subject (NO!), and there has already been a few polls on discord, what really bothers me is the lack of uniformity in the doc and examples, so I would like to have a decision that we can document and refer to.

So... here it is, a totally fair poll. Also, @cart vote probably counts around 9000:
- 👍 : no `_system` suffix
- 👎 : yup, I need a `_system` suffix to know my system is a system because indicating a function type by it's name is the best

**Comment by alice-i-cecile** [comment]
Regardless of the decision, I'd like to see a consistency pass on the code base for this :) It's messy, and slightly increases the friction for contributors.

**Comment by Nilirad** [comment]
I'm for `*_system` until we need a proc macro or something to discover a system in the codebase e.g. for the editor

Something like:

```
#[system]
fn do_stuff() {
   ... 
}
```

Would look neat, but it needs to be _only_ applicable to systems and should have a reason to exist.

**Reply by mockersf** [reply]
There is already a way to discover systems in the editor: we just need to list the implementors of the trait. Relying on function name for editor discovery seems... flaky

**Reply by Nilirad** [reply]
How would that be done? Reflection or using rust-analyzer?

**Reply by alice-i-cecile** [reply]
Systems can register themselves into the App when they're first run or added to the schedule :)

**Reply by Nilirad** [reply]
> Systems can register themselves into the App when they're first run or added to the schedule :)

🤔 Automatic system insertion could be interesting, but I don't get what do you actually intend here. Could you please elaborate? I was thinking more about system discoverability to visually add them to the app schedule via the editor, on a graph-like view.

**Reply by alice-i-cecile** [reply]
> I was thinking more about system discoverability to visually add them to the app schedule via the editor, on a graph-like view.

Yep, that's exactly what I had in mind. The schedule graph of [`bevy_mod_debug_dump`](https://github.com/jakobhellermann/bevy_mod_debugdump) has some preliminary functionality for this :D

It [looks like](https://github.com/jakobhellermann/bevy_mod_debugdump/blob/main/src/schedule_graph.rs) it can grab the appropriate information from the schedule already.

**Reply by Nilirad** [reply]
> There is already a way to discover systems in the editor: we just need to list the implementors of the trait.

As shown by [this comment](https://github.com/bevyengine/bevy/discussions/2804#discussioncomment-1330071), it is not the best approach though, since there could be functions that are accidentally systems but are not intended to be (e.g. functions without parameters), so, I'm back to the `#[system]` opinion 😅

EDIT: We can also get rid of the suffix this way though 😉

**Comment by TarekAS** [comment]
By not using the `_system` suffix, how would one easily tell if a function is meant to be used as a system (without having to look at the parameters)? A comment? Putting all systems in a `systems` module? A macro? A convention is needed to tell systems apart from regular functions.

Until then, the `_system` suffix makes more sense, especially to us beginners (actually I prefer the `system_` prefix 😈)

**Reply by Ixentus** [reply]
The parameters are part of the function signature. Why would you ignore them?

**Reply by TarekAS** [reply]
@Ixentus Well for starters, systems don't _need_ parameters to be valid systems, you could easily have systems without any parameter. 

Second, nothing stops a regular function from having the same parameter types as those used in a system AFAIK.

Third, and most subjectively, I find it annoying to have to read the function parameters in order to tell if it's a system or not. When quickly skimming through hundreds of lines of code I tend to look only look at the function names first, parameters later. It's less readable than simply using the system prefix/suffix.

**Reply by Nilirad** [reply]
> puts `_system` suffix on non-system functions

That's the evilest thing I've ever thought about 😝

**Reply by Nilirad** [reply]
> Second, nothing stops a regular function from having the same parameter types as those used in a system AFAIK.

Well, That's a system by definition. It doesn't matter if you `add_system` it or not, it will implement the trait in each case.

**Reply by TarekAS** [reply]
You are right, but also any function with no parameters also implement the trait, making it a system by definition (which it may or may not be in practice).

**Reply by Nilirad** [reply]
At the end we'll probably need a `#[system]` macro in the future to catch systems without false positives.

We're already going on the route of `#[derive(Component)]`, I would not be surprised if we do the same with systems.

**Comment by marko-lazic** [comment]
I'm holding all my system functions inside impl of the struct CustomPlugin. Then I'm calling them inside impl Plugin for CustomPlugin. I never have systems outside of a impl scope. That does not feel like distinct enough sometimes.

**Reply by Nilirad** [reply]
Last time I used functions defined in impl blocks (a long time ago) it didn't compile.

**Reply by marko-lazic** [reply]
Here's what it looks like throughout my project with bevy 0.5
Every system is defined in its plugin and I have plugin for almost everything :)
```
struct CombatPlugin;

impl CombatPlugin {
    fn fight() {}
}

impl Plugin for CombatPlugin {
    fn build(&self, app: &mut AppBuilder) {
        app.add_system(Self::fight.system());
    }
}
```
Maybe I should have less plugins but I like separating things into their own game logic collections. I don't like systems laying around in random files without belonging to something. I also don't like having directory or file named systems and then shoving all systems there.
I will probably go with whatever we decide. I already had situation with uniformity of naming systems and decided to go without _system for now.

**Reply by alice-i-cecile** [reply]
Oh that's interesting, to store the systems as methods on plugin... Makes sense I suppose! I wonder if this is a pattern we should consider adopting, especially if you used a glob `from Self import::*;` statement in the Plugin impl.

I'm generally pretty happy with small plugins in general, I tend to find them easier to reason about and they're helpful for organizing code.

**Comment by antmelnyk** [comment]
_system prefix feels kinda forced and not natural. System is basically just a function, defined by its arguments. I think division by system/non-system functions should be done on the structural level, let's say functions in one mod, real systems in another.

**Reply by Nilirad** [reply]
If I would have a serious project, I would definitely use that approach. But this type of code organization is overkill for smaller things like Bevy examples.

**Comment by themaxhero** [comment]
I think calling ```*_system.system()``` sounds redundant.
Probably a convention to have each system on it's own file would solve this problem of having to find out what kind of function is it.
But also, correct me if I'm wrong. But, if you look at the function signature it's kinda hard to not know what is that function about.
That remembers me the practice of using b prefix on boolean, s prefix on strings, and things like that.
